### DISCUSSION TITLE
9.10a1 General Purpose Blocks Feedback

### DISCUSSION THREAD

**Discussion by facelessuser** [facelessuser/pymdown-extensions#1868] [Show and tell]
Documentation: https://facelessuser.github.io/pymdown-extensions/extensions/blocks/

The purpose of this discussion is to allow feedback for the latest iteration of General Purpose Blocks 9.10.a1. It should be noted that any and all feedback is welcome, but there is a focus on feedback specifically related to syntax. In particular, we have two main points we are hoping to get feedback on, but again, any feedback on syntax is welcome.

Currently, we have two general options that are only available during the alpha (and maybe beta) stage. By the final release, we'd like to remove these options with a permanent solution.

1. Which is preferred, syntax with `:::` or `///`? Currently, the option `colon_syntax` allows selecting between the current default syntax of `///` and the colon syntax (`:::`). Which is generally preferred by the community?

2. The new general purpose Blocks extension allows YAML configs in block headers to set per block options. YAML fences are currently optional and can be omitted, but using `require_yaml_fences` can make them mandatory.

    The thinking is that some people may find including the YAML fences tedious, so allowing them to be optional allows a user not use them, but this requires us to enforce a new line between block headers and block content, headers in this sense meaning the start of the block and the config.

    ```
    /// name
    option: value

    content requires a new line ^here^
    ///

    /// name
    
    even when there is no config, we still need a new line ^here^
    ///
    ```

    Requiring YAML fences means we can easily identify when there is a config, so no new line is required. This means a user can still add a new line if desired, but if not desired, they can omit it.

    ```
    /// name
    ---
    option: value
    ---
    content can use a new line ^here^ or omit it
    ///

    /// name
    no config, no new line needed
    ///
    ```

**Comment by waylan** [comment]
Very Cool! I read through the docs [here][1]. Is that the current up-to-date version? Of course, a rendered version would be better, but I understand why you wouldn't want to override the current stable version docs.

Regarding the two requested points of feedback...

1. I prefer `///`. 👍 from me.
2. I'm indifferent on this one. I see advantages to both. I say you get to paint the shed as you built it. While you could leave the config option in place, I don't think it makes sense to have two valid (but incompatible)  syntax forms out there. I understand that being able to play with both during the alpha/beta stage is helpful, but I agree that a choice should be made before final release with only one option. Maybe if I get some time later, I'll install and play with it.

[1]: https://github.com/facelessuser/pymdown-extensions/pull/1777/files

**Reply by facelessuser** [reply]
> Is that the current up-to-date version? Of course, a rendered version would be better, but I understand why you wouldn't want to override the current stable version docs.

I've updated the opening post with a documentation link. Yes, it is rendered. I just didn't think to post the link here.

>  I understand that being able to play with both during the alpha/beta stage is helpful, but I agree that a choice should be made before final release with only one option. Maybe if I get some time later, I'll install and play with it.

Yep, I do not want to support two different methods. I'm mainly looking for good arguments to sway me one way or the other. I do have my personal leanings, and worst case, I'll just go with that. This just gives me a chance to hopefully understand people's largest concerns before I commit to a final solution.

Assuming we don't have any amazing ideas that cause me to rethink the syntax, the next step will be to formalize the API and document it. During that period, I'll be more focused on making sure there is sufficient flexibility to give people what they need, or handle things I just didn't think of.

I think we are on our way to a reasonable generic block format.

**Comment by 2bndy5** [comment]
Upon reading the [tab docs](https://facelessuser.github.io/pymdown-extensions/extensions/blocks/tab), I noticed that tab 2 in the "new tabbed group" example links to the previous example's tab 2 (instead of switching to the tab).

**Reply by facelessuser** [reply]
Oh, there is a reason for this. The examples are rendered in a sandbox, so they have no awareness of each other. I should explicitly make them have different IDs because of this.

**Reply by facelessuser** [reply]
I pushed an update to the docs so that these examples work as expected. It should probably be live as soon as GitHub refreshes them.

**Reply by 2bndy5** [reply]
Cool, that fixed it, thanks.

I'm also seeing that the tab 1 contents aren't displayed on first page load:
![image](https://user-images.githubusercontent.com/14963867/205376624-aa706901-214c-478e-bb6d-f06b76ec1220.png)

After switching to tab 2 and back to tab 1, the contents are displayed (even after page refresh). This behavior is consistent across Firefox and Chrome. I had to clear the site data in my browser to reproduce the screenshot above.

**Reply by facelessuser** [reply]
I'll look into it. It's supposed to function just like the Tabbed extension (it's meant to replace it in the future), but maybe there are a couple of bugs to shake out still. I'll take a closer look.

**Reply by facelessuser** [reply]
You know, I think it is specifically because I'm running the example in a sand box. A page will generate each tab set with a unique ID, but since they are both rendered in a sandbox, they get the same starting ID for the tab set. The main docs are still using the Tabbed extension, but I'm trying to demonstrate the new tab functionality in these sandboxes, so conflicts. I think I'll just render the result using Tabbed outside of a sandbox as the output is identical. That'll fix the ID issues.

**Reply by facelessuser** [reply]
Once I switch over to using the new general purpose blocks for the entire documentation, this won't be an issue anymore.

**Reply by 2bndy5** [reply]
Sorry to "be that guy". I'm just super analytical...

**Reply by facelessuser** [reply]
No worries, this was a good catch. I pushed some sane rendering now.

I hadn't realized that rendering tabs examples in sandboxes would cause this issue, but IDs need to be unique across the page, and if they aren't, well, you see what happens. I also have to run the general blocks in a sandbox for rendering as I don't have them loaded by default for the site yet, so they are done on demand in examples.

Docs should be fixed once GitHub refreshes.  The examples get the point across now and work as expected. This is a non-issue once I formally use the new general-purpose blocks site-wide.

**Reply by 2bndy5** [reply]
Yep, that fixed it! Thanks for being so incredibly responsive!

**Comment by squidfunk** [comment]
One quick thing I immediately noticed – the front matter separator `---` induces a `h2` element for __when the extension is not enabled or supported__, e.g. on GitHub Pages. This leads to the problem that the element is even listed in the table of contents.

__Example__

``` markdown
# Lorem ipsum

Lorem ipsum dolor sit amet.

/// details | Some summary
---
type: warning
---

Some content
///
```

__Result__

![screenshot-localhost-8000-1670078134957](https://user-images.githubusercontent.com/932156/205446253-50cb314d-e8e9-4398-98e8-198495bbaab9.png)

**Reply by facelessuser** [reply]
So this brings up some good questions:

1. Do we care?
2. If we do care:
    -  Do we want a non-standard YAML fence instead?
    - Do we want to skip fences and instead require each line to "comment" out each line of the config, `/ key: value` (or some other convention)?
    - Do we want to just keep fences optional and enforce the new line before content?

**Reply by facelessuser** [reply]
An alternative, more Markdown compatible form of giving options could be something like this:

```
/// name | argument
| key1: value
| key2: value

content
///
```

Then, if a user really cares about not triggering syntax in a non-Python Markdown scenario, it should work fine.

/// name | argument
| key1: value
| key2: value

content
///

**Reply by facelessuser** [reply]
I think doing this could allow people to use YAML fences if they just don't care about how it renders in something like GitHub, but if they do, they can use this inline form instead. Both are easily identifiable and can allow us to tell where config ends and content starts.

**Reply by facelessuser** [reply]
It may simply be that using YAML fences, in general, is not a good idea. Looking at the original Myst inspiration, they never show their Markdown friendly `:::` syntax using YAML fences either, they only show that when using code block syntax, and instead use an escaping methodology `: key: value`. Only in their code block alternative form do they use YAML fences.

**Reply by squidfunk** [reply]
I guess giving users the option to opt-out of this behavior would be great, but as I understand that's already possible with `require_yaml_fences` and omitting them, right? Sorry for my uninformedness, I'm still trying to catch up with the recent developments yet trying to be helpful.

**Reply by squidfunk** [reply]
Looping in @oprypin, who's advocating for a nicer rendering of non-standard Markdown on GitHub (e.g. via [markdown-callouts](https://oprypin.github.io/markdown-callouts/)).

**Reply by facelessuser** [reply]
The idea is to settle on a reasonable syntax, so the options are added to help people determine what seems reasonable, but such options are not meant to persist in the final solution.

I had not bothered to try and render this syntax in GitHub, but in retrospect, YAML syntax, using traditional fences, probably isn't going to work well. Even if we allow the fences to be omitted, and require a new line after the config, you could still have the YAML config have portions rendered as indented code blocks if options are nested deep enough.

We either need to have any content directly coming after the header with no new line treated as a config OR we require some token for marking content after a header to instruct it that the content is a config. If you happen to have a conflict, you just enforce it with a new line.

While we don't have to use `|`, I will use it as a placeholder for now:

```
/// name | argument
| key1: value
| key2: value

New line is not required after config, but can if desired
///

/// name | argument
no config, and no new line required
///
```

or 

```
/// name | argument
key1: value
key2: value

New line is required before content
///

/// name | argument

New line is required before content even if there is no config
///
```

**Reply by squidfunk** [reply]
Personally, I think the required newline will trip up more users than any syntax we settle on, because it might easily be forgotten. However, that's just my first guess. I haven't used it enough to confidently decide on what's better. Thanks for putting out the alpha! I think having a longer alpha phase, or at least considering this feature experimental for some time is a good idea to get early adopters to start using it while retaining the flexibility to change things when we run into problems.

**Reply by facelessuser** [reply]
Yeah, I don't want to rush it.

Also, my personal feelings align with yours as far as the new line is concerned.

I think in the second alpha I'll drop the YAML fences and provide this more explicit per line token for testing.

**Reply by 2bndy5** [reply]
> While we don't have to use `|`, I will use it as a placeholder for now

What about a char that reads (to the uninformed) like lines that relate to the block start?
```markdown
/// note | title
/key1: value
/key2: value

content
///
```

I'm still indifferent about the requirement for a blank line. Coming from rST, I'm accustomed to directives that mandate a blank line if the directive accepts arguments or if a directive option(s) are specified.

<details><summary>rST example</summary>

```rst
.. admontion:: required title as arg that
    can span multiple consecutive lines in doc src
    :class: special-admonition-css-class

    Requires blank line.

.. note::
    Accepts no args and no option provided, so no need for blank line.

.. note::
    :class: special-admonition-css-class

    Required blank line because directive option was specified.
```
</details>

**Reply by 2bndy5** [reply]
> I think the required newline will trip up more users than any syntax we settle on, because it might easily be forgotten.

Ideally, there would be a warning from the parser that implies that no content was provided (typically because the content was mistaken for config by the parser). This would be best generated if the block options are known to the parser, and any unrecognized options (like intended content) would be output in the warning.

**Reply by facelessuser** [reply]
> What about a char that reads (to the uninformed) like lines that relate to the block start?

This is fine as well, I have no problem using `/`.

> I'm still indifferent about the requirement for a blank line. Coming from rST, I'm accustomed to directives that mandate a blank line if the directive accepts arguments or if a directive option(s) are specified.

It's not about when the config is present. The requirement is annoying because it is required when there is no config. But, if we require something like `/` in front of every line of the config, the requirement goes away.

> Ideally, there would be a warning from the parser that implies that no content was provided (typically because the content was mistaken for config by the parser). This would be best generated if the block options are known to the parser, and any unrecognized options (like intended content) would be output in the warning.

Block options that are not allowed are already rejected. Block options that use the wrong type will also be rejected. The config has to come first before we create the block, if the options are incorrect or do not follow the spec of the block, the block will not be created.
It is possible we could add warnings, but no other Markdown plugins do this, and you would not have the resolution to turn off just these warnings in something like MkDocs, but if we added warnings, maybe we could make it an explicit option or something. I won't worry about warnings right now though.

So, for consistency, I'll likely require configs to follow something like this, but the new line does not need to be enforced if we are doing this.  We could require it for config only, but there is no parser need, so I'll likely let the user do it if they like to.

```
/// note | title
/key1: value
/key2: value

content
///
```

**Reply by waylan** [reply]
I have to admit I hadn't given much thought to how the syntax would render in standard Markdown either. The YAML delimitators are definitely a problem there. That is one of the benefits of using indented blocks. In standard Markdown you just get a code block with the raw meta-data displayed in its original form. 

But, of course, we are trying to avoid indented blocks. So then we need to find a solution that does have a reasonable fallback. However, I very much dislike the suggestions made in this thread. Prefixing every line is not something I have ever liked. As an example, consider Markdown's blockquotes. Of course, you can just prefix the first line of a paragraph, but if you have ever edited a paragraph with hard line breaks and the prefix on every line, it gets tedious real fast. Additionally, in my opinion, the YAML data should be able to be copy/pasted without alteration. Requiring a prefix on every line makes that impossible.

I would much prefer the already implemented blank line syntax. Or possibly alternate delimitators (use different characters which Markdown doesn't see an HRs).

**Reply by facelessuser** [reply]
I'm open to suggestions, the blank line doesn't fix YAML code getting ugly in renders and possibly syntax highlighters if you have nested YAML configs that exceed 4 spaces.

**Reply by facelessuser** [reply]
One other thing I will note is that Python Markdown, when things are very reliant on new lines, has always generated questions now and then as users get tripped up. This is true with HTML support and other areas. If not having to strictly require a new line between config and content helps me avoid answering the same question over and over, not expressly requiring new lines would be nice This will definitely not prevent people from using new lines between config and content.

This doesn't, of course, solve the issue with having to require blockquote style line syntax for configs. Personally, I wish Markdown never had the indented code block rule as I feel fenced code blocks are better, and we have to workaround indentation if we want to not have confusing results in other parsers and editors.

Obviously, inline syntax can still trigger in code blocks, but I guess it would be less grievous.

Alternatively, I guess you could really lean into code blocks and require config to be indented.

```
/// name | argument
    key2: value
    key2: value

content
///
```

---

/// name | argument
    key2: value
    key2: value

content
///

---

Most Markdown parsers and/or editors should not treat the YAML config as code in this case, or will treat it all as code, which will at least make things consistent. In this case, we could enforce a newline after the config, but if no config is present, no new line is needed unless a block specific requires it, for instance, you would definitely need a new line to have an indented code block:

```
/// name | argument

    code
    block
///
```

**Reply by facelessuser** [reply]
Yeah, it seems if you indent the entire config, parsers (and I assume editors) will either treat it as part of the header block or as a code block.

https://babelmark.github.io/?text=%2F%2F%2F+name+%7C+argument%0A++++key1%3A+value%0A++++key2%3A+value%0A++++++key3%3A+value%0A++++key4%3A+value%0A%0Acontent%0A%2F%2F%2F

If you have some sort of blockquote style leading token, they all pretty much treat the config the same way.

https://babelmark.github.io/?text=%2F%2F%2F+name+%7C+argument%0A%2F+key1%3A+value%0A%2F+key2%3A+value%0A%2F+++key3%3A+value%0A%2F+key4%3A+value%0A%0Acontent%0A%2F%2F%2F

There's obviously no way to prevent inline parsing from mucking it up, but I think at least some attempt to keep syntax highlighted and rendering in editors in a somewhat reasonable way is preferrable.

I can't say it makes much of a difference to me whether we indent the content or add a blockquote style toke to extend the header as they basically equate to the same thing for me, but it's possible someone has some idea more outside the box.

**Reply by waylan** [reply]
I really like indenting the YAML (hadn't considered that option before). And I think the two outcomes by other parsers are both reasonable. You could allow (but not require) blank lines surrounding the YAML for those who will want it to display as a code block on GitHub etc. In other words, the following examples would all render the same by the extension:

```
/// name
    option: value
content with no blank lines at all
///
```

```
/// name
    option: value

content separated by a newline
///
```

```
/// name

    option: value

content and config separated by newlines
///
```

Then users who care can use the third style and will get nicely rendered code blocks of the YAML configs for their blocks on other sites/tools.

**Reply by facelessuser** [reply]
The one edit I would make to your statement is that we would not allow the following for configs, or better put, the config will not work and will simply be a content code block:

```
/// name

    option: value

content and config separated by newlines
///
```

This is because we'd like to create the element when we get the header, that is why we need the header to contain the config, or maybe put more simply, the header, which is received as a single block, must completely describe the block. The content can be added lazily. There are times, like in lists and such, that you can't look ahead in blocks to see if you can find the config, so requiring the first block to describe the block and provide the config is currently essential. Maybe in the future we can figure out how to relax this, but starting out, I would not allow the above.

**Reply by waylan** [reply]
For me, that kills the attractiveness of the option as users can't guarantee the config will be rendered as a code block in other tools/sites. Even so, an intended config is still the best compromise we have yet, but without the above, its not as great.

Another option might be to use an alternate character for YAML delimitators. For example, `...` is already in the [spec][1] for the end of a YAML block. What if we used that:

```
/// name
...
option: value
...
content
///
```

That [doesn't conflict with any other Markdown syntax][2]. Although, I think I like the indented option better.

[1]: https://yaml.org/spec/1.0/#id2489959
[2]: https://babelmark.github.io/?text=%2F%2F%2F+name%0A...%0Aoption%3A+value%0A...%0Acontent%0A%2F%2F%2F

**Reply by facelessuser** [reply]
I did think about an alternate YAML config fence, and that is certainly a fine option. There is one small issue, and it could be one that we just accept if we feel it's not worth the effort to prevent.

If you have deeply nested options (I can't predict exactly how blocks will be used, so it is certainly possible to have more complex configs), a syntax highlighter or editor _might_ translate a deeply nested config into a code block:

```markdown
/// name
...
test:
  test:
    test: value
...
content
///
```

Babelmark indicated a couple of cases where this happens:

```html
<p>
  /// name ... test: test:
</p>
<div>
  <pre>
    <code>
      test: value

    </code></pre>
</div>
<p>
  ...
</p>
<p>
  content ///
</p>
```

Do we care enough? Personally, I care more about my syntax highlighter not confusing me more than I care about some random parser not being able to handle what we are doing. Python Markdown and CommonMark seem to handle this fine.  FWIW the `...` fence seems to not cause the syntax highlighters I use to go crazy as long as there are no blank new lines in the config, which we forbid anyways as the header must be one, complete block.

So, this is certainly viable if we are willing to accept that there may be some highlighters and/or parsers that may treat a deeply nested YAML config portion as a code block. I think these may be rare though.

**Reply by facelessuser** [reply]
I guess there is another failing of YAML fences. Options that use a list:

```
/// name
...
test:
- this
- that
...
*content*
///
```

---

/// name
...
test:
- this
- that
...
*content*
///

**Reply by facelessuser** [reply]
Unfortunately, the best way to mitigate such nonsense is some sort of per-line token, either indentation or a character. Particularly if we are trying to mitigate things like GitHub rendering to make it at least somewhat readable on the repo.

**Reply by waylan** [reply]
One of the benefits of prefixing every line of the config is that the entire config gets treated the same by other tools/parsers. An added benefit of using indentation as the prefix is that it is easier to work with in editors as they have built-in support for it (with the exception of web forms). So, if the concern is how other tools/parsers treat the blocks, then indentation is the best option IMO. However, if the desire is ease to create, then prefixes are not so good (including indentation when working in web forms). In that case, full line delimitators are better, whether those be blank lines or lines of repeating characters. Its a matter of priorities. 

As YAML list items will get parsed as Markdown list items by other parsers, I don't see any reason to be concerned about the delimitators being parsed as HRs. In other words, if we use delimitators, then lets just stick with YAML's defaults (`---`). However, if that is undesirable, then per-line prefixes is the only reasonable way to avoid that, which means indentation is the way to go IMO.

**Reply by facelessuser** [reply]
I realize that we cannot fully prevent Markdown parsers and syntax highlighters from accidentally parsing or highlighting something in the config that we don't intend it to, but I do think the fence solution allows for far more egregious issues. I do realize now that switching to `...` from `---` only provides very minimal improvements, so if we stick with YAML fences, `---` is really not that much worse.

The intent is not to completely be compatible with the existing Markdown. The big thing with general blocks is to:

1. Make creating special blocks easy and consistent so we don't have to invent various new syntaxes with all the related issues in doing that.
2. To prevent deeply nested indentation which can currently happen with admonitions, lists, and other constructs.
3. To minimize how painful it is for people to write using the special Markdown in their editors.
4. To a lesser extent, it would be nice to minimize the pain of looking at a non-standard Markdown in a GitHub repo, etc.

So, with all of that in mind, I think some sort of per-line delimiter makes sense. I am more than happy to adopt indentation as it will not propagate deeply nested constructs as it is only a syntax to continue the header of these blocks to define per-block options. It will be interesting to get opinions from the larger audience, but I think I'm okay with an indentation for config only. Worst case, if the majority of people really hate indentation, a special line character can be considered. This may be one of these things where whatever solution we settle on will make someone unhappy, but I'm to start with something and see where the conversation goes.

I think in alpha 2, I'll remove the YAML fence and the related option and try out indented configs and get feedback.

**Reply by ofek** [reply]
> per-line delimiter

Example?

**Reply by facelessuser** [reply]
@ofek It would be something like what is shown below. You extend the header with some token:

```
/// name
/ key1: value
/ key2: value

content
///
```

**Reply by facelessuser** [reply]
When a place YAML fences in the bin, I may add a new alpha option to try out indentation vs line delimiter. That may help people get a feel for the two options so they can give feedback.

**Comment by vokimon** [comment]
I am biased towards how currently  [`customblocks`](https://vokimon.github.io/markdown-customblocks/) works, you know. So, take these proposals with a grain of salt.

## Block marker

You ask for opinions. Well, I feel like `:::` is nicer as `///` is a common sequence in code comments. But no strong opinion here. Pandoc and R markdown use `:::` for custom blocks. It is a similar use case but not exactly the same one, they just generate divs you can populate with attributes with the `{}` syntax. I use `///` in the examples bellow just to use the current default.

## Key parameters

Fenced or unfenced YAML? Neither. I do love YAML, but i feel that having any named/optional parameter specified as yaml metadata in every block type is too verbose. Even more verbose if we finally use fences for that. I feel like key parameters could be better solved inline, like:

```markdown
/// blocktype key=val key2="val with spaces"
content
///
```

For long parameter lists, still we could give the option to split them in lines by using, for example, the line continuation escaping, which is a common idiom in many languages (not currently supported in `customblocks` but in the TODO list):
```markdown
/// blocktype \
  key=val \
  key2="val with spaces"
content
///
```

I know that, by removing yaml parameters, we left an scenario uncovered: having structured data. But, i feel that they will be the 2% or less of the cases, and we are messing everybody with verbose syntax.  If your block type really needs structured data, let block-type authors consider the yaml part of their content. We could even provide them a helper (say `extract_metadata(content, fenced=true)` ) for extension authors to use it if they need it. But, for simpler cases (the 98%), it could be resolved with a single line and not being that verbose.

## Positional parameters

Regarding the positional parameters, for me the `|` is spurious. Whatever comes after the first word indicating the type, should be a parameter. To provide more than one parameter, I like bash like parameter passing is simple and a common ground: Split by words and quote multiword values.

Both, backslash line continuation and word parameter separation with quoting for spaces are common syntax patterns that many people are used to. Indeed yaml use them.

## Flag parameters

Taking as reference what you can do with [`customblocks`](https://vokimon.github.io/markdown-customblocks/), I also miss flags (`flag`, `noflag`) which provide a natural talk like interface. An example:

```markdown
/// figure myimage.jpg width=200px nolightbox thumb
This is the **caption**
///
```

- `myimage.jpg` is a positional parameter, the obvious parameter
- `width` is a regular key parameter
- `nolightbox` gives false to  `lightbox` flag
- `thumb` gives true to `thumb` flag

## Single line closing

Whenever you don't need any content the block gets kind of ugly.

```markdown
/// figure myimage.jpg width=200px nolightbox thumb
///
```

An idea could be having single line closing if you do not need any content:

```markdown
/// figure myimage.jpg width=200px nolightbox thumb ///
```

Of course, if content is to be indented neither option would be necessary.

## Escaping closing tag

Also because the content is at the same indentation level, we could end up having some content containing `///`.
Indeed whenever you have an example of the extension syntax.
I guess you solved it as the examples should be documented somehow but i couldn't find the examples.


Just some ideas.

**Reply by facelessuser** [reply]
> You ask for opinions. Well, I feel like ::: is nicer as /// is a common sequence in code comments. But no strong opinion here. Pandoc and R markdown use ::: for custom blocks.

Noted. I am aware of others using `:::`, so I can understand this bias. Generally, I'm seeing how many people really, really care about this vs the ones indifferent and/or like `///`.

> I feel like key parameters could be better solved inline

Currently, I'm against this. The reason is that we reserve the inline argument for common things like titles in abominations etc. It allows a more simple format for things the user is most interested in. Allowing inline config complicates the parsing and the user input if you are wanting to do something like adding a Markdown title with Markdown syntax. For instance, in Admonitions, it can be a Markdown title that will be parsed by Markdown to add emphasis or any other feature supported by other inline plugins.

If we start using inline key-value pairs, it greatly complicates the syntax and requires escaping Markdown and all sorts of issues. With that said, any generic block can treat this inline argument however they want, so if you want a plugin that uses that argument as a sequence of key-value pairs or space-separated parameters, that is up to the plugin author.

Granted, the nature of a general block may be more complex or less complex than others. I can envision certain blocks heavily using the config as opposed to the inline attribute depending on what the block needs to accomplish.

> For long parameter lists, still we could give the option to split them in lines by using, for example, the line continuation escaping, which is a common idiom in many languages (not currently supported in customblocks but in the TODO list):

I am pretty strongly against this approach. Markdown is not like other languages and this will not prevent syntax highlighters and GitHub renders from greatly misinterpreting configs as a  mess.   I don't think Markdown has to mimic things like C/C++, JavaScript, or other programming languages in this regard. This also conflicts with the EscapeAll plugin that I support which allows you to add `<br>` tags when you escape a line ending.

> Flag Parameters

This is dependent on the idea of inline parameters which I'm currently not a fan of, but a plugin could certainly repurpose inline arguments to do this.

> Single line closing

This also assumes we are using inline configs, so I'm currently not considering this.

> Escaping closing tag

I'm not sure I quite understand., but if the content includes `///` that is not meant to be processed as a block start or end, we could simply allow `/` to be escaped `\/`. I don't think I've added `/` to the escaped character list yet, so that should probably be done.

**Reply by 2bndy5** [reply]
>> Flag Parameters
>
>This is dependent on the idea of inline parameters which I'm currently not a fan of, but a plugin could certainly repurpose inline arguments to do this.

I imagine this could also be implemented (by plugin authors) with multi-lined configs keys that have no value assigned.

---
I'm starting to see that `:::` may be preferred (if only a throw back to existing syntax). It would probably make migrating between parsers more smooth (from the user standpoint).

**Reply by facelessuser** [reply]
> I'm starting to see that ::: may be preferred (if only a throw back to existing syntax). It would probably make migrating between parsers more smooth (from the user standpoint).

That may be how people generally feel. Keep in mind though, that even if we somewhat mirror what some people are used to, we have no intention of forcing our implementation to match others out there, and there are multiple different ones that utilize `:::` but do it differently.

**Reply by waylan** [reply]
I prefer `///` to `:::` for at least two reasons.

1. It makes it clear that the syntax is different than (and not compatible with) the preexisting solutions. Using `:::` could actual cause some users to think the syntax is compatible and then they will be surprised when things don't work the same. It is actually doing the users a service to not use `:::` and make it clear it is different.
2. I find `:::` to be much more busy to look at that `///`. I realize this is subjective and others may disagree, but I find `///` to feel much cleaner.

My subjective opinions aside, I think the first is reason enough to support using something different than `:::`. Of the various proposals made `///` was chosen. So unless you can propose something other than `:::` which is clearly better, I would stick with `///`. But, I'm not building the shed so I don't get to paint it.

However, there is a concern that occurred to me while reading @2bndy5's feedback: how are we handling nested blocks? For example, can a user nest a `details` block inside an `admonition` block? That is easy when the block content of indented, but less clear then it is not as is the case here. I know that fended code blocks can use a different number of backticks. Does that work with blocks (use `////` for example)? I don't see it mentioned in the docs.

**Reply by ofek** [reply]
I also prefer `///`

**Reply by facelessuser** [reply]
> However, there is a concern that occurred to me while reading @2bndy5's feedback: how are we handling nested blocks? For example, can a user nest a `details` block inside an `admonition` block? That is easy when the block content of indented, but less clear then it is not as is the case here. I know that fended code blocks can use a different number of backticks. Does that work with blocks (use `////` for example)? I don't see it mentioned in the docs.

Nesting is not a problem.

```
//// note | Title
Some content.

/// details | Summary
Some details
///
////
```

While there is no technical reason to have the incrementing `///` for nesting, it helps visually, especially with lots of nesting.

@waylan, as far as your first point, I generally agree, which is why I went further to explicitly state I have no intentions of explicitly matching some other implementation. I don't mind taking inspiration, but I don't want to bind our implementation to another. I do think using `///` makes it explicitly clear that yes, while there are similarities, this is different.

**Reply by waylan** [reply]
Regarding the nesting, I had a vague recollection that we had discussed that earlier in the planning stage, but as there was no mention of it in the documentation, I wasn't sure if it was implemented. IIRC, you wanted to require that the longer delimitators must be on the outer block, not the inner one. That is not something that is consistent with fenced code blocks (which allow either way) and so needs to be explicitly documented IMO.

**Reply by facelessuser** [reply]
Is that really the case? If it is, I was not aware. I will need to re-evaluate. It is a general requirement in a number of Markdown parsers I am aware of, so I assumed it would be in Python Markdown as well. As far as SuperFences is concerned, I think it is a requirement, so it is at least consistent with that plugin, which I maintain. See below in GitHub, reversing does not work.

`````
```
````
this
````
```
`````

EDIT: had to remove the visual rendering as it was pretty broken.

Regardless, I will document it for clarity.

**Reply by waylan** [reply]
As a reminder, Python-Markdown and PHP Markdown Extra were the very first implementations of Fenced Code Blocks. Michelf and and I came up with the syntax together (with input from others) and implemented it. We never discussed or implemented such a restriction. It would seem that the restriction was added later by other implementations. Personally, I was not aware that such a restriction was in place anywhere until you brought it to my attention. I still maintain that such a restriction should not exist.

**Reply by facelessuser** [reply]
While I do understand this now, I will have to consider the implications of relaxing this constraint in things I support. As far as things are right now, the outer fence of both Blocks and SuperFences hold to this restriction. I won't promise a relaxing of this, only that I'll look into it and consider this change as a possibility. 

It is quite likely that this is a painless change, but I'll have to consider it carefully. I do imagine there is a likely possibility of breakage.

**Reply by facelessuser** [reply]
Okay, I was wrong, SuperFences does allow larger fences under smaller, the only requirement is that they are different. So, it would be reasonable to match that in Blocks.

<img width="795" alt="Screenshot 2022-12-09 at 10 17 28 AM" src="https://user-images.githubusercontent.com/1055125/206757216-ffbccbd2-965c-4436-a038-ce0888c06870.png">

My syntax highlighter hates it though, so I imagine this might be why I never use fences in this direction:

<img width="478" alt="Screenshot 2022-12-09 at 10 18 07 AM" src="https://user-images.githubusercontent.com/1055125/206757290-e15ed740-76e0-4692-ab60-628ac471467b.png">

**Reply by facelessuser** [reply]
Next alpha will allow nesting of any length of fences as long as they differ from the current one. While the example below only shows escaping YAML config with `/`, there will be an alpha option to use indentation, I just haven't added it yet. There will also be some minor nesting fixes

<img width="785" alt="Screenshot 2022-12-09 at 11 44 33 AM" src="https://user-images.githubusercontent.com/1055125/206771535-9ac2b03f-e2e7-47f1-89f3-8d2d6730e110.png">

**Comment by facelessuser** [comment]
Discussion is moving to the new Alpha 2 discussion thread: https://github.com/facelessuser/pymdown-extensions/discussions/1883. Please use the new thread for further discussions. I will be releasing the alpha 2 shortly.
